src/cpp/core/FileLock.cpp

```diff
@@ -15,163 +15,144 @@
+   
+   FilePath locksConfPath(kLocksConfPath);
+   if (!locksConfPath.exists())
+      return;
+   
+   std::map<std::string, std::string> conf;
+   Error error = core::readStringMapFromFile(locksConfPath, &conf);
+   if (error)
+      LOG_ERROR(error);
+   
+   if (conf.count("lock-type"))
+   {
+      std::string lockType = to_lower_copy(conf["lock-type"]);
+      if (lockType == "advisory")
       {
-         lock.unlock();
-         return false;
+         FileLock::setLockType(FileLock::LOCKTYPE_ADVISORY);
+      }
+      else if (lockType == "link-based")
+      {
+         FileLock::setLockType(FileLock::LOCKTYPE_LINKBASED);
       }
       else
       {
-         return true;
+         std::string msg = std::string() +
+               "unrecognized lock type '" + lockType + "'";
+         LOG_ERROR_MESSAGE(msg);
       }
    }
-   catch(boost::interprocess::interprocess_exception& e)
+   
+   if (conf.count("lock-refresh-rate"))
    {
-      Error error(boost::interprocess::ec_from_exception(e), ERROR_LOCATION);
-      error.addProperty("lock-file", lockFilePath);
-      LOG_ERROR(error);
-      return false;
+      long refreshRate = ::atol(conf["lock-refresh-rate"].c_str());
+      if (refreshRate == 0)
+         LOG_ERROR_MESSAGE("invalid lock refresh rate '" + conf["lock-refresh-rate"] + "'");
+      else
+         FileLock::setLockRefreshRate(boost::posix_time::seconds(refreshRate));
+   }
+   
+   if (conf.count("lock-timeout-interval"))
+   {
+      long timeoutInterval = ::atol(conf["lock-timeout-interval"].c_str());
+      if (timeoutInterval == 0)
+         LOG_ERROR_MESSAGE("invalid lock timeout interval '" + conf["lock-timeout-interval"] + "'");
+      else
+         FileLock::setLockTimeoutInterval(boost::posix_time::seconds(timeoutInterval));
    }
 }
 
+} // namespace locks
 
-struct FileLock::Impl
-{
-   FilePath lockFilePath;
-   boost::interprocess::file_lock lock;
-};
+// default values for static members
+FileLock::LockType FileLock::type_(FileLock::LOCKTYPE_ADVISORY);
+boost::posix_time::seconds FileLock::timeoutInterval_(30);
+boost::posix_time::seconds FileLock::refreshRate_(20);
 
-FileLock::FileLock()
-   : pImpl_(new Impl())
+boost::shared_ptr<FileLock> FileLock::create()
 {
+   switch (FileLock::getLockType())
+   {
+   case LOCKTYPE_ADVISORY:  return boost::shared_ptr<FileLock>(new AdvisoryFileLock());
+   case LOCKTYPE_LINKBASED: return boost::shared_ptr<FileLock>(new LinkBasedFileLock());
+   }
 }
 
-FileLock::~FileLock()
+void FileLock::refresh()
 {
+   AdvisoryFileLock::refresh();
+   LinkBasedFileLock::refresh();
 }
 
-Error FileLock::acquire(const FilePath& lockFilePath)
-{
-   using namespace boost::interprocess;
-
-   // make sure the lock file exists
-   if (!lockFilePath.exists())
-   {
-      Error error = core::writeStringToFile(lockFilePath, "");
-      if (error)
-         return error;
-   }
+namespace {
 
-   // try to acquire the lock
+void schedulePeriodicExecution(
```

Comment:
You need to make sure the error_code argument is passed in the bind and check for errors in the implementation. See https://github.com/rstudio/rstudio/blob/d5ab5dfa53c057a289179826a94622b715146e79/src/cpp/core/include/core/http/AsyncServerImpl.hpp#L399-L427